doc/adr/0007-support-reference-inputs-in-constraint-library.rst

```diff
@@ -0,0 +1,47 @@
+.. _support_reference_inputs_in_constraint_library:
+
+ADR 7: Support reference inputs in constraint library
+=====================================================
+
+Date: 2022-08-09
+
+Authors
+-------
+
+koslambrou <konstantinos.lambrou@iohk.io>
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+After the Vasil HF, the Cardano blockchain will support `reference` inputs by adding a new field in the transaction data type.
+With reference inputs, transactions can take a look at UTXOs without actually spending them.
+
+Thus, we need to adapt our transaction constraint data type (`TxConstraints`) to support referencing UTXOs.
+
+Decision
+--------
+
+* We will add the data constuctor `MustReferenceOutput TxOutRef` to the `TxConstraints` data type.
+
+* The PlutusV1 on-chain implementation of this new constraint will simply return `False`.
+  However, `cardano-ledger` throws a phase-1 validation error if transactions that use the some of the new features (reference inputs, inline datums and reference scripts) try to execute PlutusV1 scripts.
+  See the `Babbage era ledger specification <https://hydra.iohk.io/job/Cardano/cardano-ledger/specs.babbage-ledger/latest/download-by-type/doc-pdf/babbage-changes>`_.
+  Therefore, this on-chain implementation should **never** result in a phase-2 validation error.
```

Comment:
What does it mean that it will never result in a phase-2 validation error? It's important that it will be caught during the first phase?

Reply:
Basically, if the user tries to submit a transaction which:
- uses reference inputs, inline datums or reference scripts
- includes a V1 scripts

This will result in a phase-1 error by `cardano-ledger` - meaning that the transaction will fail before being submitted in the blockchain. If it was a phase-2 error, it means that the transaction will be part of the blockchain and will be failing (meaning collateral inputs will be used).

Reply:
Oh, didn't know about the validation phases, thanks!